modules/apps/expando/expando-web/src/main/resources/META-INF/resources/js/GeoLocationField.js

```diff
@@ -0,0 +1,56 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import MapBase from '@liferay/map-common/js/MapBase.es';
+
+export default function ({inputName, mapName}) {
+	const geolocationField = {
+		init() {
+			MapBase.get(mapName, (map) => {
+				map.on(
+					'positionChange',
+					geolocationField.onPositionChange,
+					geolocationField
+				);
+			});
+		},
+		onPositionChange(event) {
+			const inputNode = document.querySelector(
+				'[name="' + inputName + '"]'
+			);
+
+			const location = event.newVal.location;
+
+			if (inputNode) {
+				inputNode.setAttribute(
+					'value',
+					JSON.stringify({
+						latitude: location.lat,
+						longitude: location.lng,
+					})
+				);
+			}
+
+			const locationNode = document.getElementById(
+				inputName + 'Location'
+			);
+
+			if (locationNode) {
+				locationNode.innerHTML = event.newVal.address;
+			}
+		},
+	};
+
+	geolocationField.init();
```

Comment:
Used in the context of a `liferay-frontend:component`, this will be invoked with `new GeolocationField(...)`, so even though this should work, we're basically expecting classes with `constructor`  and `destroy`/`dispose` methods to handle the component lifecycle.

We have a bit of documentation in https://github.com/liferay/liferay-frontend-projects/blob/master/guidelines/dxp/liferay_frontend_component.md which we might want to expand.

See https://github.com/liferay-frontend/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib/src/main/java/com/liferay/frontend/taglib/servlet/taglib/ComponentTag.java#L202 for the implementation details.

Reply:
Hmm, I noticed we had a handful of examples in DXP that export functions that return an object with a `dispose` method.

Personally, I think it would be better to stay away from declaring classes in the JS world unless they are totally necessary.

Reply:
I'd be okay with that, I only think a "component" contract should be clear and not depend on side-effects. One thing we've discussed in the past is to add some other abstraction like `liferay-frontend:script` where people can just load scripts for their side effects.

When using `liferay-frontend:component` we're doing:

``` 
const component = new MyComponent.default(data);

Liferay.component(componentId, component);
```

So, based on this, we're calling this with `new` and I'd expect we're registering some `undefined` component which is what new might be returning here. This could lead to other sort of problems down the road, just based on the expectations about what `liferay-frontend:component` is doing. 

Again, if we don't want to promote class-based components that's totally fine, but I think we should provide a better, clearer way to invoke those.